File: dotnet-accounting-app/AccountStore.cs
using System.Globalization;
using System.Text;
using System.Text.RegularExpressions;

namespace Accounting;

internal sealed partial class AccountStore
{
    private const long OpeningBalanceCents = 100_000;
    private readonly string file = Path.GetFullPath(
        Environment.GetEnvironmentVariable("ACCOUNT_STORE") ?? "build/dotnet-account.cents");
    private string ParentDirectory => Path.GetDirectoryName(file)
        ?? throw new InvalidOperationException("The account store must have a parent directory.");

    [GeneratedRegex(@"\A(?:0|[1-9][0-9]{0,7})\n?\z", RegexOptions.CultureInvariant)]
    private static partial Regex StoredCentsPattern();

    internal long ReadBalance()
    {
        string text;
        try
        {
            using var input = File.OpenRead(file);
            // Read one byte beyond the eight-digit, optional-LF contract without BOM sniffing.
            Span<byte> bytes = stackalloc byte[10];
            var count = input.ReadAtLeast(bytes, bytes.Length, throwOnEndOfStream: false);
            if (count > 9)
            {
                throw new InvalidDataException("Account store exceeds the 9-byte UTF-8 contract.");
            }
            text = new UTF8Encoding(false, true).GetString(bytes[..count]);
        }
        catch (FileNotFoundException)
        {
            return OpeningBalanceCents;
        }
        catch (DirectoryNotFoundException)
        {
            // A file blocking an ancestor path is not a missing, new account.
            Directory.CreateDirectory(ParentDirectory);
            return OpeningBalanceCents;
        }
        // L-09: only an absent store is new; corruption must never reset the account.
        if (!StoredCentsPattern().IsMatch(text))
        {
            throw new InvalidDataException("Account store must contain integer cents between 0 and 99999999.");
        }
        return long.Parse(text.TrimEnd('\n'), CultureInfo.InvariantCulture);
    }

    internal void WriteBalance(long cents)
    {
        Money.RequireBalance(cents);
        var directory = ParentDirectory;
        Directory.CreateDirectory(directory);
        var temporary = Path.Combine(directory, $".account-{Guid.NewGuid():N}.tmp");
        try
        {
            File.WriteAllText(temporary, cents.ToString(CultureInfo.InvariantCulture) + "\n",
                new UTF8Encoding(false, true));
            File.Move(temporary, file, overwrite: true);
        }
        finally
        {
            File.Delete(temporary);
        }
    }
}

File: java-accounting-app/src/AccountStore.java
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.charset.CodingErrorAction;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.NoSuchFileException;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.util.regex.Pattern;

final class AccountStore {
    private static final long OPENING_BALANCE_CENTS = 100_000;
    private static final Pattern STORED_CENTS = Pattern.compile("(?:0|[1-9][0-9]{0,7})\\n?");
    private final Path file;

    AccountStore() {
        String configured = System.getenv("ACCOUNT_STORE");
        file = Path.of(configured == null ? "build/java-account.cents" : configured).toAbsolutePath();
    }

    long readBalance() throws IOException {
        byte[] bytes;
        try (var input = Files.newInputStream(file)) {
            // Eight digits plus optional LF; the extra byte detects oversized files.
            bytes = input.readNBytes(10);
        } catch (NoSuchFileException missing) {
            return OPENING_BALANCE_CENTS;
        }
        if (bytes.length > 9) {
            throw new IOException("account store exceeds the 9-byte UTF-8 contract");
        }
        String text = StandardCharsets.UTF_8.newDecoder()
            .onMalformedInput(CodingErrorAction.REPORT)
            .onUnmappableCharacter(CodingErrorAction.REPORT)
            .decode(ByteBuffer.wrap(bytes)).toString();
        // L-09: only an absent store is new; corruption must never reset the account.
        if (!STORED_CENTS.matcher(text).matches()) {
            throw new IOException("account store must contain integer cents between 0 and 99999999");
        }
        return Long.parseLong(text.strip());
    }

    void writeBalance(long cents) throws IOException {
        Money.requireBalance(cents);
        Files.createDirectories(file.getParent());
        Path temporary = Files.createTempFile(file.getParent(), ".account-", ".tmp");
        try {
            Files.writeString(temporary, cents + "\n", StandardCharsets.UTF_8);
            Files.move(temporary, file, StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING);
        } finally {
            Files.deleteIfExists(temporary);
        }
    }
}
